src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -267,15 +178,58 @@ public final boolean isActive() {
 
     @Nonnull
     private List<FlowNode> loadParents(List<String> parentIds) {
-        List<FlowNode> _parents = new ArrayList<>(parentIds.size());
-        for (String parentId : parentIds) {
+        List<FlowNode> _parents;
+
+        if (parentIds.size() == 1) {
             try {
-                _parents.add(exec.getNode(parentId));
+                return Collections.singletonList(exec.getNode(parentIds.get(0)));
             } catch (IOException x) {
                 LOGGER.log(Level.WARNING, "failed to load parents of " + id, x);
+                return Collections.emptyList();
+            }
+        } else {
+            _parents = new ArrayList<>(parentIds.size());
+            for (String parentId : parentIds) {
+                try {
+                    _parents.add(exec.getNode(parentId));
+                } catch (IOException x) {
+                    LOGGER.log(Level.WARNING, "failed to load parents of " + id, x);
+                }
             }
+            return _parents;
+        }
+    }
+
+    /**
+     * Get the first enclosing block ID for this node. Can be null.
+     * @return
+     */
+    @CheckForNull
+    public String getEnclosingId() {
+        FlowNode enclosing = this.exec.findEnclosingBlockStart(this);
+        return enclosing != null ? enclosing.getId() : null;
+    }
+
+    /**
+     * Get the list of enclosing blocks, starting from innermost, for this node. May be empty. Loads the flow nodes into
+     * the transient {@code enclosingBlocks} field if needed.
+     */
+    @Nonnull
+    public List<FlowNode> getEnclosingBlocks() {
```

Comment:
*faints with happiness*

Reply:
Might want to go to a doctor if that makes you faint ;)

Reply:
What can I say, I'm fond of easy APIs. =)